airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py

```diff
@@ -0,0 +1,605 @@
+        # We also have to check that the headings are in the breaking changes dict
+        ordered_breaking_changes = sorted(breaking_changes.keys(), reverse=True)
+        ordered_expected_headings = [f"{expected_version_header_start}{version}" for version in ordered_breaking_changes]
+
+        ordered_heading_versions = []
+        for line in migration_guide_content.splitlines():
+            stripped_line = line.strip()
+            if stripped_line.startswith(expected_version_header_start):
+                version = stripped_line.replace(expected_version_header_start, "")
+                ordered_heading_versions.append(version)
+
+        if ordered_breaking_changes != ordered_heading_versions:
+            return self.create_check_result(
+                connector=connector,
+                passed=False,
+                message=textwrap.dedent(
+                    f"""
+                Migration guide file for {connector.name_from_metadata} has incorrect version headings.
+                Check for missing, extra, or misordered headings, or headers with typos.
+                Expected headings: {ordered_expected_headings}
+                """
+                ),
+            )
+        return self.create_check_result(
+            connector=connector,
+            passed=True,
+            message="The migration guide is correctly templated",
+        )
+
+
+class CheckDocumentationExists(DocumentationCheck):
+    name = "Connectors must have user facing documentation"
+    description = (
+        "The user facing connector documentation should be stored under `./docs/integrations/<connector-type>s/<connector-name>.md`."
+    )
+
+    def _run(self, connector: Connector) -> CheckResult:
+        if not connector.documentation_file_path or not connector.documentation_file_path.exists():
+            return self.fail(
+                connector=connector,
+                message="User facing documentation file is missing. Please create it under ./docs/integrations/<connector-type>s/<connector-name>.md",
+            )
+        return self.pass_(
+            connector=connector,
+            message=f"User facing documentation file {connector.documentation_file_path} exists",
+        )
+
+
+class CheckDocumentationContent(DocumentationCheck):
+    """
+    For now, we check documentation structure for sources with sl >= 300.
+    """
+
+    applies_to_connector_languages = [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]
+    applies_to_connector_ab_internal_sl = 300
+    applies_to_connector_types = ["source"]
+
+
+class CheckDocumentationLinks(CheckDocumentationContent):
+    name = "Links used in connector documentation are valid"
+    description = f"The user facing connector documentation should update invalid links in connector documentation."
+    valid_status_codes = [200, 403, 401, 405, 429, 503]  # we skip 4xx due to needed access
```

Comment:
I don't get how >=400 status code can be valid...
I think the only valid status code are 200 and redirects.

Reply:
Test performs general unauthorized request, so 401 and 403 response for links that requires auth. I added these statuses because validation of links in some cases requires permission, auth or right method, that we don't know while testing docs. So this statuses can be skipped. Overall we are only interested in 404 not found.